extensions/azurePublishNew/src/node/processStatusTracker.ts

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+import { ProcessStatus } from '@botframework-composer/types';
+import { v4 as uuid } from 'uuid';
+
+type ProcessStatusStart = Omit<Partial<ProcessStatus>, 'id' | 'projectId' | 'processName' | 'log'> & {
+  projectId: string;
+  processName: string;
+};
+
+type ProcessStatusUpdate = Pick<Partial<ProcessStatus>, 'status' | 'message' | 'config'>;
+
+/**
+ * Creates a tracker of ProcessStatus.
+ * The tracker avoids throwing errors to prevent status tracking from halting the process itself.
+ */
+export const createProcessStatusTracker = () => {
+  const statusById: Record<string, ProcessStatus> = {};
+
+  /**
+   * Starts tracking a new process
+   */
+  const start = (startStatus: ProcessStatusStart): ProcessStatus => {
+    const id = uuid();
+    statusById[id] = {
+      message: '',
+      status: 202,
+      time: new Date(),
+      ...startStatus,
+      id: id,
+      log: startStatus.message ? [startStatus.message] : [],
+    };
+    return statusById[id];
+  };
+
+  /**
+   * Gets the latest status of the process with the specified ID.
+   */
+  const get = (id: string) => {
+    return statusById[id];
+  };
+
+  /**
+   * Gets the first, most recent matching process name.
+   */
+  const getByProcessName = (processName: string) => {
+    const matches = Object.values(statusById)
+      .filter((status) => status.processName === processName)
+      .sort((a, b) => b.time.valueOf() - a.time.valueOf());
+    return matches?.[0];
+  };
+
+  /**
+   * Gets the first, most recent matching process name.
+   */
+  const getByProjectId = (projectId: string) => {
+    const matches = Object.values(statusById)
+      .filter((status) => status.projectId === projectId)
+      .sort((a, b) => b.time.valueOf() - a.time.valueOf());
+    return matches?.[0];
+  };
+
+  /**
+   * Updates the status of the process with the specified ID.
+   */
+  const update = (id: string, update: ProcessStatusUpdate) => {
```

Comment:
update as arg name is the same as the function name, maybe use a different name?

Reply:
isn't the update obj partial? should it check if it has any value before it overrides existing values with potential undefined?